File: src/mnemonic/builder/font_fetcher.py
"""Koruriフォント取得機能

KoruriフォントをGitHubからダウンロードし、キャッシュに保存する機能を提供する。
Koruriフォントは Apache License 2.0 でライセンスされている。
"""

from __future__ import annotations

from dataclasses import dataclass
from pathlib import Path
from typing import TYPE_CHECKING

import httpx

from mnemonic.cache import get_cache_dir

if TYPE_CHECKING:
    pass


class FontDownloadError(Exception):
    """フォントダウンロードに関する例外"""

    pass


@dataclass(frozen=True)
class FontInfo:
    """フォント情報を表す不変オブジェクト

    Attributes:
        name: フォント名（例: Koruri-Regular）
        path: フォントファイルのパス
        version: フォントのバージョン
    """

    name: str
    path: Path
    version: str


# Koruriフォントのダウンロード設定
# GitHubリリースにはZIPファイルが存在しないため、リポジトリから直接TTFをダウンロード
KORURI_VERSION = "master"
KORURI_FONT_NAME = "Koruri-Regular"
KORURI_TTF_FILENAME = f"{KORURI_FONT_NAME}.ttf"
KORURI_DOWNLOAD_URL = (
    f"https://raw.githubusercontent.com/Koruri/Koruri/master/{KORURI_TTF_FILENAME}"
)


def get_font_cache_dir() -> Path:
    """フォントキャッシュディレクトリを取得する

    Returns:
        フォントキャッシュディレクトリのパス
    """
    return get_cache_dir() / "fonts"


class FontFetcher:
    """Koruriフォントを取得するクラス

    GitHubからKoruriフォントをダウンロードし、
    ローカルキャッシュに保存する機能を提供する。

    Example:
        >>> fetcher = FontFetcher()
        >>> font_info = await fetcher.get_font()
        >>> print(font_info.path)
        /home/user/.cache/mnemonic/fonts/Koruri-Regular.ttf
    """

    DEFAULT_TIMEOUT = 60.0  # HTTPリクエストのタイムアウト秒数

    def __init__(
        self,
        cache_dir: Path | None = None,
        http_client: httpx.AsyncClient | None = None,
    ) -> None:
        """FontFetcherを初期化する

        Args:
            cache_dir: フォントのキャッシュディレクトリ。
                       Noneの場合はデフォルトのキャッシュディレクトリを使用。
            http_client: HTTPクライアント（テスト用の依存性注入）。
                         Noneの場合は内部でクライアントを作成。
        """
        self._cache_dir = cache_dir or get_font_cache_dir()
        self._http_client = http_client
        self._owns_client = http_client is None

    async def get_font(self) -> FontInfo:
        """フォントを取得する

        キャッシュに有効なフォントが存在する場合はそれを返し、
        存在しない場合はダウンロードする。

        Returns:
            フォント情報

        Raises:
            FontDownloadError: ダウンロードに失敗した場合
        """
        if self.is_cache_valid():
            cached_path = self.get_cached_font_path()
            if cached_path is not None:
                return FontInfo(
                    name=KORURI_FONT_NAME,
                    path=cached_path,
                    version=KORURI_VERSION,
                )

        return await self.download_font()

    async def download_font(self) -> FontInfo:
        """フォントをダウンロードする

        GitHubからKoruri-Regular.ttfを直接ダウンロードしてキャッシュに保存する。

        Returns:
            フォント情報

        Raises:
            FontDownloadError: ダウンロードに失敗した場合
        """
        self._cache_dir.mkdir(parents=True, exist_ok=True)

        # TTFを直接ダウンロード
        ttf_content = await self._download_ttf()

        # TTFファイルを保存
        font_path = self._cache_dir / KORURI_TTF_FILENAME
        font_path.write_bytes(ttf_content)

        return FontInfo(
            name=KORURI_FONT_NAME,
            path=font_path,
            version=KORURI_VERSION,
        )

    async def _download_ttf(self) -> bytes:
        """Koruri-Regular.ttfをダウンロードする

        Returns:
            TTFファイルのバイナリデータ

        Raises:
            FontDownloadError: ダウンロードに失敗した場合
        """
        client = self._http_client
        close_client = False

        if client is None:
            client = httpx.AsyncClient(timeout=self.DEFAULT_TIMEOUT)
            close_client = True

        try:
            response = await client.get(KORURI_DOWNLOAD_URL, follow_redirects=True)
            response.raise_for_status()
            return response.content
        except httpx.RequestError as e:
            raise FontDownloadError(f"ネットワークエラー: {e}") from e
        except httpx.HTTPStatusError as e:
            raise FontDownloadError(f"HTTPエラー {e.response.status_code}: {e}") from e
        finally:
            if close_client and client is not None:
                await client.aclose()

    def is_cache_valid(self) -> bool:
        """キャッシュが有効かどうかを確認する

        Returns:
            キャッシュが存在し有効な場合はTrue
        """
        font_path = self._cache_dir / KORURI_TTF_FILENAME
        return font_path.exists()

    def get_cached_font_path(self) -> Path | None:
        """キャッシュされたフォントのパスを取得する

        Returns:
            フォントファイルのパス。キャッシュが無い場合はNone。
        """
        font_path = self._cache_dir / KORURI_TTF_FILENAME
        if font_path.exists():
            return font_path
        return None

File: internal/builder/font_fetcher.go
package builder

import (
	"errors"
	"fmt"
	"io"
	"net/http"
	"os"
	"path/filepath"
	"time"

	"github.com/na2na-p/mnemonic/internal/cache"
)

// ErrFontDownload はKoruriフォントのダウンロードに関する基本エラー。
var ErrFontDownload = errors.New("フォントのダウンロードに失敗しました")

// Koruriフォント関連の定数。
const (
	// KoruriVersion はKoruriフォントのバージョン識別子。
	//
	// why not: GitHubリリースにはZIPアセットが存在しないため、リポジトリのmasterブランチ
	// から直接TTFを取得する（Python版と同じ制約）。そのためバージョンはブランチ名に留まる。
	KoruriVersion = "master"
	// KoruriFontName はフォント名。
	KoruriFontName = "Koruri-Regular"
	// KoruriTTFFilename はTTFファイル名。
	KoruriTTFFilename = KoruriFontName + ".ttf"
)

// defaultKoruriDownloadURL は既定のKoruriフォントダウンロードURL。
const defaultKoruriDownloadURL = "https://raw.githubusercontent.com/Koruri/Koruri/master/" + KoruriTTFFilename

// fontDownloadTimeout はフォントダウンロードの既定タイムアウト秒数。
const fontDownloadTimeout = 60 * time.Second

// FontInfo はフォント情報を表す不変値。
type FontInfo struct {
	Name    string
	Path    string
	Version string
}

// FontCacheDir はフォントキャッシュディレクトリを返す。
func FontCacheDir() (string, error) {
	dir, err := cache.Dir()
	if err != nil {
		return "", err
	}

	return filepath.Join(dir, "fonts"), nil
}

// FontFetcher はKoruriフォントをGitHubからダウンロードし、
// ローカルキャッシュに保存する。
type FontFetcher struct {
	// CacheDir はフォントの保存先ディレクトリ。空文字列の場合はFontCacheDir()を使用する。
	CacheDir string
	// HTTPClient はHTTPリクエストに使用するクライアント。
	HTTPClient *http.Client
	// SourceURL はダウンロード元URL。
	//
	// why not: httptestサーバーに差し替えてテストするための注入口
	// （TemplateDownloader.APIBaseURLと同じ方針）。空文字列の場合は
	// defaultKoruriDownloadURLを使用する。
	SourceURL string
}

// NewFontFetcher はFontFetcherを初期化する。
// httpClientがnilの場合はfontDownloadTimeoutを持つ既定クライアントを使用する。
func NewFontFetcher(cacheDir string, httpClient *http.Client) *FontFetcher {
	if httpClient == nil {
		httpClient = &http.Client{Timeout: fontDownloadTimeout}
	}

	return &FontFetcher{CacheDir: cacheDir, HTTPClient: httpClient}
}

// httpClient はHTTPClientフィールドを返す。nilの場合（FontFetcher{}のゼロ値を
// NewFontFetcherを介さず直接構築した場合）はfontDownloadTimeoutを持つクライアントを
// 都度生成して返し、nilポインタ参照によるpanicを避ける。
func (f *FontFetcher) httpClient() *http.Client {
	if f.HTTPClient != nil {
		return f.HTTPClient
	}

	return &http.Client{Timeout: fontDownloadTimeout}
}

func (f *FontFetcher) sourceURL() string {
	if f.SourceURL != "" {
		return f.SourceURL
	}

	return defaultKoruriDownloadURL
}

func (f *FontFetcher) cacheDir() (string, error) {
	if f.CacheDir != "" {
		return f.CacheDir, nil
	}

	return FontCacheDir()
}

// GetFont はフォントを取得する。
// キャッシュに有効なフォントが存在する場合はそれを返し、存在しない場合はダウンロードする。
func (f *FontFetcher) GetFont() (FontInfo, error) {
	if f.IsCacheValid() {
		if path, ok := f.GetCachedFontPath(); ok {
			return FontInfo{Name: KoruriFontName, Path: path, Version: KoruriVersion}, nil
		}
	}

	return f.DownloadFont()
}

// DownloadFont はGitHubからKoruri-Regular.ttfを直接ダウンロードしてキャッシュに保存する。
func (f *FontFetcher) DownloadFont() (FontInfo, error) {
	dir, err := f.cacheDir()
	if err != nil {
		return FontInfo{}, err
	}

	if err := os.MkdirAll(dir, 0o750); err != nil {
		return FontInfo{}, fmt.Errorf("%w: キャッシュディレクトリの作成に失敗しました: %w", ErrFontDownload, err)
	}

	content, err := f.downloadTTF()
	if err != nil {
		return FontInfo{}, err
	}

	fontPath := filepath.Join(dir, KoruriTTFFilename)
	if err := os.WriteFile(fontPath, content, 0o600); err != nil {
		return FontInfo{}, fmt.Errorf("%w: フォントファイルの保存に失敗しました: %w", ErrFontDownload, err)
	}

	return FontInfo{Name: KoruriFontName, Path: fontPath, Version: KoruriVersion}, nil
}

func (f *FontFetcher) downloadTTF() ([]byte, error) {
	req, err := http.NewRequest(http.MethodGet, f.sourceURL(), nil)
	if err != nil {
		return nil, fmt.Errorf("%w: リクエストの構築に失敗しました: %w", ErrFontDownload, err)
	}

	resp, err := f.httpClient().Do(req)
	if err != nil {
		return nil, fmt.Errorf("%w: ネットワークエラー: %w", ErrFontDownload, err)
	}
	defer func() { _ = resp.Body.Close() }()

	if resp.StatusCode >= 300 {
		return nil, fmt.Errorf("%w: HTTPエラー %d", ErrFontDownload, resp.StatusCode)
	}

	content, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("%w: レスポンスの読み込みに失敗しました: %w", ErrFontDownload, err)
	}

	return content, nil
}

// IsCacheValid はキャッシュが有効かどうかを確認する。
func (f *FontFetcher) IsCacheValid() bool {
	dir, err := f.cacheDir()
	if err != nil {
		return false
	}

	_, err = os.Stat(filepath.Join(dir, KoruriTTFFilename))

	return err == nil
}

// GetCachedFontPath はキャッシュされたフォントのパスを取得する。
// キャッシュが無い場合はok=falseを返す。
func (f *FontFetcher) GetCachedFontPath() (string, bool) {
	dir, err := f.cacheDir()
	if err != nil {
		return "", false
	}

	path := filepath.Join(dir, KoruriTTFFilename)
	if _, err := os.Stat(path); err != nil {
		return "", false
	}

	return path, true
}
